### PR TITLE
C++: Summary metrics queries

### PR DESCRIPTION
**Pull request by dbartol**
Resolves https://github.com/github/codeql-c-analysis-team/issues/239

This is a first attempt at implementing, for C++, the set of summary queries that we expect all languages to implement to help diagnose extraction failures and build configuration problems. See the spec in [this document](https://docs.google.com/document/d/1V3zpkj0OGh8GEUVwACRx7fiafE5zklujAftZaYUyf9s/edit?usp=sharing). The five queries are:

- Total number of source files (including .c/.cpp and header files)
- Total number of lines of text across all source files
- Total number of lines of code across all source files
- Number of lines of text in each source file
- Number of lines of code in each source file

I've added some simple unit tests that cover all five of these.

### CODE REVIEW

File: cpp/ql/src/Diagnostics/LinesOfCode.ql
Comment (geoffw0) [inline cpp/ql/src/Diagnostics/LinesOfCode.ql:10]: For the benefit of anyone else who is wondering, `getNumberOfLinesOfCode()` doesn't count blank lines and lines that only contain comments.

**Review by geoffw0 — COMMENTED** [review]
Is there a difference in meaning between a "text file" and "source file" in the PR description?  If there is, does it affect CPP?

Queries LGTM.

File: cpp/ql/src/Diagnostics/Files.ql
Comment (jbj) [inline cpp/ql/src/Diagnostics/Files.ql:5]: These queries are in a `Diagnostics` directory, and I think that describes them well. Therefore I'd write `diagnostics` or `diagnostic` rather than `metrics` in the query ids. This of course has to be coordinated across languages. What do you think, @adityasharad ?

File: cpp/ql/src/Diagnostics/Files.ql
Comment (jbj) [inline cpp/ql/src/Diagnostics/Files.ql:4]: Is `@kind metric` all the metadata we need for Code Scanning to present the results of these queries in the right place? I note that `@kind metric` is already in use by LGTM, and [the LGTM documentation](https://help.semmle.com/lgtm-enterprise/user/help/writing-custom-queries.html) specifies that such queries have two additional mandatory properties. We have existing `@kind metric` queries for Python and JavaScript in this repo.

File: cpp/ql/src/Diagnostics/Files.ql
Comment (adityasharad) [reply cpp/ql/src/Diagnostics/Files.ql:5]: In our design for extractor diagnostics we used the kind `diagnostic`, so it's up to you whether you want to give the same directory name and ID prefix for metrics+diagnostics, or you want to distinguish between those two. The IDs will be visible in the SARIF.

I have a slight preference for distinguishing them but honestly do not feel strongly about this.

File: cpp/ql/src/Diagnostics/Files.ql
Comment (adityasharad) [reply cpp/ql/src/Diagnostics/Files.ql:4]: Oh well spotted. I reused the existing `@kind metric` deliberately, but will look into making those properties optional in the toolchain. If we're not adding these queries to the LGTM suites then we don't need to make LGTM handle them.

**Comment by dbartol** [comment]
>Is there a difference in meaning between a "text file" and "source file" in the PR description? If there is, does it affect CPP?
s/text file/source file/
(I've updated the description)

File: cpp/ql/src/Diagnostics/Files.ql
Comment (asgerf) [reply cpp/ql/src/Diagnostics/Files.ql:4]: > Is `@kind metric` all the metadata we need for Code Scanning to present the results of these queries in the right place?

It sounds like we'll need a tag and/or a query suite to identify the metrics to use in Code Scanning. Any decision on what to use here, or should we just add it later when the rest of the toolchain is further along?

Comment (jbj) [reply cpp/ql/src/Diagnostics/Files.ql:4]: I can see the drawbacks of reusing `@kind metric`: we have to change the rules for additional properties and possibly change how the suites are produced. What are the advantages?

Comment (adityasharad) [reply cpp/ql/src/Diagnostics/Files.ql:4]: - Go with just `@kind metric` for now, I think that's enough for you to proceed :)
- The contents of the Code Scanning suites are determined by https://github.com/github/codeql/blob/main/misc/suite-helpers/code-scanning-selectors.yml. Regardless of what name we use here, we'll have to update that config to ship these new types of queries to Code Scanning. The suite format gives us the ability to include by `kind`.
- The LGTM suite generation script only includes queries that also have `metricType`.
- The existing meta-metric queries for JS and Python can easily be excluded by filtering out `@tags meta`, or just editing their metadata if those teams are happy with it (LGTM and Code Scanning don't include them anyway).
- I think metric is the simplest name for this property, and it gives scope to reuse the metric/treemap queries we already have that fit this purpose. That said, the choice shouldn't block you here: if we encounter trouble making the toolchain work, or need to do a final find-replace with a new name, that's on my team to fix.

Comment (asgerf) [reply cpp/ql/src/Diagnostics/Files.ql:4]: > The existing meta-metric queries for JS and Python can easily be excluded by filtering out `@tags meta`

I'd be uncomfortable with a query suite defined as exclusion-filter, i.e. "everything except queries with tags x, y, z". A simple typo in a tag could cause a query to be picked up by accident, making its way into production.

Would you please consider an inclusion-filter instead, for example,
```yaml
- include:
    kind:
    - metric
    tags contain:
    - diagnostic
```

(sorry for hijacking the thread, it seemed related to this)

File: cpp/ql/src/Diagnostics/Files.ql
Comment (dbartol) [reply cpp/ql/src/Diagnostics/Files.ql:5]: @adityasharad Is the audience for extractor diagnostics different from the audience for lines-of-code metrics? Aren't they both for determining whether CodeQL is extracting the right stuff? I suppose that they're two different levels of diagnostic (look at the lines-of-code summary, and if that looks wrong, look at extractor diagnostics).

I don't like using "metrics", because to me that implies the sort of metrics that ODASA used to do (Halstead, Cyclomatic Complexity, etc.). How about using something like `@id cpp/summary/lines-of-code` for the queries implemented here?

Comment (adityasharad) [reply cpp/ql/src/Diagnostics/Files.ql:5]: Sounds good!

File: cpp/ql/src/Diagnostics/Files.ql
Comment (jbj) [reply cpp/ql/src/Diagnostics/Files.ql]: I don't know the technicalities of how big of an effort it would be to invent a new _kind_ vs. the effort of changing the docs. I also don't know our overall philosophy of how much we try to reuse existing _kinds_ vs. create new ones.

If we do reuse the _kind_, I'm very much in support of Asger's proposal to let the suites be determined by inclusion filters rather than exclusion filters. It seems like too much of a footgun to require every future LGTM-style `metric` query to add `@tags meta`. Authors of such queries will be testing the LGTM-style functionality, and they won't notice that Code Scanning suddenly starts including extra diagnostic data in its logs because omitted a tag.

Comment (dbartol) [reply cpp/ql/src/Diagnostics/Files.ql]: Since we're already going with "summary" for the path name and `@id`, should we use `summary` as the tag, rather than `diagnostic`?

Comment (adityasharad) [reply cpp/ql/src/Diagnostics/Files.ql]: Inclusion filter sounds sensible. I think `summary` is a good tag name for now since we eventually plan to have `diagnostic` as a separate kind (for extractor diagnostics), so let's go with that. If you want both of those tag names supported later it's easy to add.

**Comment by criemen** [comment]
@jbj per discussion with @AlonaHlobina , this PR now contains only the first LoC summary metric we want to measure.
This metric query returns a LoC-count that is indicative of the database size.

Another metric query, that corresponds to a LoC-count that is relatable to the user, is postponed until further clarification from product.

Here, LoC means "newline in any C/C++/header file".

This is ready for review now.

**Comment by criemen** [comment]
Please hold back on merging, there's some news about what a LoC actually consists of, I might need to change this query again after all.

**Review by jbj — APPROVED** [review]
Wonderfully simple. LGTM.

**Comment by jbj** [comment]
Okay, holding back on merging.

**Comment by criemen** [comment]
@jbj: Ready for re-review. I guess we should also get :eyes: from the docs team?

**Comment by jbj** [comment]
@github/docs-content-codeql, please review the docs (name and description) of this new query.

File: cpp/ql/src/Summary/LinesOfCode.ql
Comment (jbj) [inline cpp/ql/src/Summary/LinesOfCode.ql:3]:
```suggestion
 * @name Total lines of C/C++ code in the database
```

[The name should not include a full stop](https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md#query-name-name).

File: cpp/ql/src/Summary/LinesOfCode.ql
Comment (jbj) [inline cpp/ql/src/Summary/LinesOfCode.ql:4]:
```suggestion
 * @description The total number of lines of C/C++ code across all files, including system headers, libraries, and auto-generated files. This is a useful metric of the size of a database. Lines of code are all lines in a file that was seen during the build that contain code, i.e. are not whitespace or comments.
```

We like using the Oxford comma.

The second sentence has some ambiguity issues around "the build that contain code", but I'll leave it to the docs team to clear that up. We used to avoid "i.e." under the Semmle style guide, but I don't know what we do these days as GitHub.

**Review by jbj — APPROVED** [review]

File: cpp/ql/src/Summary/LinesOfCode.ql
Comment (shati-patel) [inline cpp/ql/src/Summary/LinesOfCode.ql:4]:
```suggestion
 * @description The total number of lines of C/C++ code across all files, including system headers, libraries, and auto-generated files. This is a useful metric of the size of a database. For all files that were seen during the build, this query counts the lines of code, excluding whitespace or comments.
```

The final sentence was a bit ambiguous, so I've made a small suggestion. Feel free to edit if it's still unclear!

**Review by shati-patel — COMMENTED** [review]
Just a small comment about the description 😊

**Comment by criemen** [comment]
@jbj can you re-approve?

**Review by jbj — APPROVED** [review]

**Comment by asgerf** [comment]
Why did the PR checks not fail for this PR?

In the [JS version of this query](https://github.com/github/codeql/pull/5304) we get:
```
[WARN] ql/javascript/ql/src/Summary/LinesOfCode.ql: Missing required property @metricType.
[ERROR] ql/javascript/ql/src/Summary/LinesOfCode.ql: expected result pattern(s) are not present for query kind "metric" (Expected metric result pattern.)
```

**Comment by criemen** [comment]
Presumably because you run the odasa-based `check_queries` target in your CI job, whereas C++ uses a CodeQL CLI based `codeql_check_queries` target.
Try changing `check_queries(` to `codeql_check_queries(` in `build` approx. line 153.

**Pull request merged**
